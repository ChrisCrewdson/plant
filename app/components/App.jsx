const React = require('react');
const actions = require('../actions');
const PropTypes = require('prop-types');

class App extends React.Component {
  static contextTypes = {
    store: PropTypes.object.isRequired,
  };

  componentWillMount() {
    const { store } = this.context;
    const users = store.getState().get('users');
    if (!users || users.size === 0) {
      store.dispatch(actions.loadUsersRequest());
    }

    const locations = store.getState().get('locations');
    if (!locations || locations.size === 0) {
      store.dispatch(actions.loadLocationsRequest());
    }

    // const user = store.getState().get('user');
    // if(user && user.get('isLoggedIn', false)) {
    //   const locationIds = user.get('locationIds');
    //   if(!locationIds) {
    //     const userId = user.get('_id');
    //     store.dispatch(actions.loadLocationsRequest(userId));
    //   }
    // }
  }

  render() {
    return (
      <div className="react-root">
        {this.props.children}
      </div>
    );
  }
}

App.propTypes = {
  // eslint-disable-next-line react/forbid-prop-types
  children: PropTypes.object.isRequired,
};

module.exports = App;
