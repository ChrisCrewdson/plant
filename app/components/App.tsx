import React from 'react';
import PropTypes from 'prop-types';
import { actionFunc } from '../actions';

export default class App extends React.Component {
  // TODO: When tsc 3.7+ is in use remove the ! to see hint text on how to change this.
  context!: PlantContext;

  static propTypes = {
    // eslint-disable-next-line react/forbid-prop-types
    children: PropTypes.object.isRequired,
  };

  static contextTypes = {
    // eslint-disable-next-line react/forbid-prop-types
    store: PropTypes.object.isRequired,
  };

  // eslint-disable-next-line camelcase
  UNSAFE_componentWillMount() {
    const { store } = this.context;
    const { users = {}, locations = {} } = store.getState();

    // TODO: This will cause a problem for a non-initialized site
    // that has zero users and zero locations as these values
    // will be 0 which is falsy.
    const usersCount = Object.keys(users).length;
    const locationsCount = Object.keys(locations).length;

    if (!usersCount) {
      store.dispatch(actionFunc.loadUsersRequest());
    }

    if (!locationsCount) {
      store.dispatch(actionFunc.loadLocationsRequest());
    }
  }

  render() {
    const { children } = this.props;
    return (
      <div className="react-root">
        {children}
      </div>
    );
  }
}
