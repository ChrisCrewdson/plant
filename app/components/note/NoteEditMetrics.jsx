const actions = require('../../actions');
const Errors = require('../common/Errors');
const Immutable = require('immutable');
const InputCombo = require('../common/InputCombo');
const React = require('react');
const Toggle = require('material-ui/Toggle').default;
const utils = require('../../libs/utils');
const PropTypes = require('prop-types');

class NoteEditMetrics extends React.PureComponent {
  constructor() {
    super();
    this.onChange = this.onChange.bind(this);
    this.renderMetric = this.renderMetric.bind(this);
    this.renderLength = this.renderLength.bind(this);
    this.renderCount = this.renderCount.bind(this);
    this.renderWeight = this.renderWeight.bind(this);
    this.renderToggle = this.renderToggle.bind(this);

    this.metricTypes = Object.freeze({
      length: this.renderLength,
      count: this.renderCount,
      weight: this.renderWeight,
      toggle: this.renderToggle,
    });
  }

  onChange(e) {
    // console.log('onChange:', e.target.name);
    const { name: inputName } = e.target;
    const interimMetrics = this.props.interimNote.get('metrics', Immutable.Map());
    const metaMetric = utils.metaMetricsGetByKey(inputName);
    const type = metaMetric.get('type');
    // based on inputName (e.g. blossom or height) we need to lookup
    // the type to determine where to pull the value. Types of toggle
    // have their values in "checked" otherwise in "value"
    // The change will be something like:
    // {metrics: { height: 23.4, blossom: true }}

    const value = type === 'toggle' ? e.target.checked : e.target.value;

    const metrics = interimMetrics.mergeDeep({
      [inputName]: value,
    });

    this.props.dispatch(actions.editNoteChange({ metrics }));
  }

  renderLength(metaMetric, value) {
    return (<InputCombo
      changeHandler={this.onChange}
      key={metaMetric.key}
      label={metaMetric.label}
      name={metaMetric.key}
      placeholder={metaMetric.placeholder}
      type="number"
      value={value || ''}
    />);
  }

  renderCount(metaMetric, value) {
    return this.renderLength(metaMetric, value);
  }

  renderWeight(metaMetric, value) {
    return this.renderLength(metaMetric, value);
  }

  renderToggle(metaMetric, value) {
    const isToggled = value === 'true' || value === true;
    return (
      <Toggle
        toggled={isToggled}
        key={metaMetric.key}
        label={metaMetric.label}
        labelPosition="left"
        name={metaMetric.key}
        onToggle={this.onChange}
        style={{ paddingLeft: '5px', maxWidth: '200px' }}
      />
    );
  }

  /**
   * @param {object} metaMetric - All the metrics available
   * @param {string} value - the value if one has been set or an empty string.
   * @returns {object} - a rendered React component to edit this type of metric
   */
  renderMetric(metaMetric, value) {
    return this.metricTypes[metaMetric.type](metaMetric, value);
  }

  render() {
    const { interimNote } = this.props;
    const metrics = interimNote.get('metrics', Immutable.Map()).toJS();
    const metaMetrics = utils.metaMetrics.toJS();

    const renderedMetrics = metaMetrics.map(metaMetric => this.renderMetric(metaMetric, metrics[metaMetric.key] || ''));

    return (
      <div style={{ textAlign: 'left' }}>
        <Errors errors={this.props.error} />
        <div>{renderedMetrics}</div>
      </div>
    );
  }
}

NoteEditMetrics.propTypes = {
  dispatch: PropTypes.func.isRequired,
  error: PropTypes.string,
  interimNote: PropTypes.shape({
    get: PropTypes.func.isRequired,
  }).isRequired,
};

NoteEditMetrics.defaultProps = {
  error: '',
};

module.exports = NoteEditMetrics;
