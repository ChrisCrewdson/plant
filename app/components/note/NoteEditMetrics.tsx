import si from 'seamless-immutable';
import React from 'react';
import Toggle from 'material-ui/Toggle';
import PropTypes from 'prop-types';

import { Dispatch } from 'redux';
import { actionFunc } from '../../actions';
import Errors from '../common/Errors';
import InputCombo from '../common/InputCombo';
import utils from '../../libs/utils';
import { PlantAction } from '../../../lib/types/redux-payloads';

// @ts-ignore - static hasn't been defined on seamless types yet.
const seamless = si.static;

interface NoteEditMetricProps {
  dispatch: Dispatch<PlantAction<any>>;
  interimNote: UiInterimNote;
  error: string;
}

export default class NoteEditMetrics extends React.PureComponent {
  props: NoteEditMetricProps;

  metricTypes: Record<string, Function>;

  static propTypes = {
    dispatch: PropTypes.func.isRequired,
    error: PropTypes.string,
    interimNote: PropTypes.shape({
      metrics: PropTypes.object,
    }).isRequired,
  };

  static defaultProps = {
    error: '',
  };

  constructor(props: NoteEditMetricProps) {
    super(props);
    this.booleanHandler = this.booleanHandler.bind(this);
    this.dispatchChange = this.dispatchChange.bind(this);
    this.onChange = this.onChange.bind(this);
    this.renderCount = this.renderCount.bind(this);
    this.renderLength = this.renderLength.bind(this);
    this.renderMetric = this.renderMetric.bind(this);
    this.renderToggle = this.renderToggle.bind(this);
    this.renderWeight = this.renderWeight.bind(this);

    this.metricTypes = Object.freeze({
      length: this.renderLength,
      count: this.renderCount,
      weight: this.renderWeight,
      toggle: this.renderToggle,
    });
    this.props = props;
  }

  /**
   * Change Handler for InputCombo
   */
  onChange(e: React.ChangeEvent<HTMLInputElement>): void {
    const { name, value } = e.target;
    this.dispatchChange(name, value);
  }

  booleanHandler = (e: React.MouseEvent<HTMLInputElement>, isInputChecked: boolean): void => {
    const { name } = e.currentTarget;
    this.dispatchChange(name, isInputChecked);
  };

  /**
   * Handle multiple change value types
   */
  dispatchChange(name: string, value: string | boolean): void {
    const { interimNote, dispatch } = /** @type {NoteEditMetricProps} */ (this.props);
    const interimMetrics = interimNote.metrics || {};

    const metrics = seamless.set(interimMetrics, name, value);

    // The change will be something like:
    // {metrics: { height: 23.4, blossom: true }}
    dispatch(actionFunc.editNoteChange({ metrics }));
  }

  renderLength(metaMetric: MetaMetric, value: any) {
    const renderValue = (value || value === 0) ? value.toString() : '';
    return (
      <InputCombo
        changeHandler={this.onChange}
        id={metaMetric.key}
        key={metaMetric.key}
        label={metaMetric.label}
        name={metaMetric.key}
        placeholder={metaMetric.placeholder}
        type="number"
        value={renderValue}
      />
    );
  }

  renderCount(metaMetric: MetaMetric, value: any) {
    return this.renderLength(metaMetric, value);
  }

  renderWeight(metaMetric: MetaMetric, value: any) {
    return this.renderLength(metaMetric, value);
  }

  renderToggle(metaMetric: MetaMetric, value: any) {
    const isToggled = value === 'true' || value === true;
    return (
      <Toggle
        key={metaMetric.key}
        label={metaMetric.label}
        labelPosition="left"
        name={metaMetric.key}
        onToggle={this.booleanHandler}
        style={{ paddingLeft: '5px', maxWidth: '200px' }}
        toggled={isToggled}
      />
    );
  }

  /**
   * @param metaMetric - All the metrics available
   * @param value - the value if one has been set or an empty string.
   * @returns - a rendered React component to edit this type of metric
   */
  renderMetric(metaMetric: MetaMetric, value: number | boolean): object {
    return this.metricTypes[metaMetric.type](metaMetric, value);
  }

  render() {
    const { interimNote, error } = this.props;
    const {
      metrics = {} as NoteMetric,
    } = interimNote || {};
    const { metaMetrics } = utils;

    const renderedMetrics = metaMetrics.map((metaMetric) => {
      const { key } = metaMetric;
      const value = metrics[key];
      // const renderValue = (value || value === 0) ? value : '';
      return this.renderMetric(metaMetric, value);
    });

    return (
      <div style={{ textAlign: 'left' }}>
        <Errors errors={error} />
        <div>
          {renderedMetrics}
        </div>
      </div>
    );
  }
}
