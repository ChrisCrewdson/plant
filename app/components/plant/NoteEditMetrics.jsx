const actions = require('../../actions');
const Errors = require('../Errors');
const Immutable = require('immutable');
const InputCombo = require('../InputCombo');
const React = require('react');
const Toggle = require('material-ui/Toggle').default;
const utils = require('../../libs/utils');

class NoteEditMetrics extends React.Component {
  constructor() {
    super();
    this.onChange = this.onChange.bind(this);
    this.renderMetric = this.renderMetric.bind(this);
    this.renderLength = this.renderLength.bind(this);
    this.renderCount = this.renderCount.bind(this);
    this.renderWeight = this.renderWeight.bind(this);
    this.renderToggle = this.renderToggle.bind(this);

    this.metricTypes = Object.freeze({
      length: this.renderLength,
      count: this.renderCount,
      weight: this.renderWeight,
      toggle: this.renderToggle,
    });
  }

  onChange(e) {
    console.log('onChange:', e.target.name);
    const {name: inputName} = e.target;
    const interimMetrics = this.props.interimNote.get('metrics', Immutable.Map());
    const type = utils.metricsGetType(inputName);
    // based on inputName (e.g. blossom or height) we need to lookup
    // the type to determine where to pull the value. Types of toggle
    // have their values in "checked" otherwise in "value"
    // The change will be something like:
    // {metrics: { height: 23.4, blossum: true }}

    const value = type === 'toggle' ? e.target.checked : e.target.value;

    const metrics = interimMetrics.mergeDeep({
      [inputName]: value
    });

    this.props.dispatch(actions.editNoteChange({metrics}));
  }

  renderLength(metaMetric, value) {
    return (<InputCombo
      changeHandler={this.onChange}
      key={metaMetric.key}
      label={metaMetric.label}
      name={metaMetric.key}
      placeholder={metaMetric.placeholder}
      value={value || ''}
    />);
  }

  renderCount(metaMetric, value) {
    return this.renderLength(metaMetric, value);
  }

  renderWeight(metaMetric, value) {
    return this.renderLength(metaMetric, value);
  }

  renderToggle(metaMetric, value) {
    const isToggled = value === 'true' || value === true;
    return (
      <Toggle
        defaultToggled={isToggled}
        key={metaMetric.key}
        label={metaMetric.label}
        labelPosition='left'
        name={metaMetric.key}
        onToggle={this.onChange}
        style={{paddingLeft: '5px', maxWidth: '150px'}}
      />
    );
  }

  /**
   * @param {object} metaMetric - All the metrics available
   * @param {string} value - the value if one has been set or an empty string.
   * @returns {object} - a rendered React component to edit this type of metric
   */
  renderMetric(metaMetric, value) {
    return this.metricTypes[metaMetric.type](metaMetric, value);
  }

  render() {
    const { interimNote } = this.props;
    const metrics = interimNote.get('metrics', Immutable.Map()).toJS();
    const metaMetrics = utils.metrics.toJS();

    const mets = metaMetrics.map((metaMetric) => {
      return this.renderMetric(metaMetric, metrics[metaMetric.key] || '');
    });

    return (
      <div style={{textAlign: 'left'}}>
        <Errors errors={this.props.error} />
        <div>{mets}</div>
      </div>
    );
  }
}

NoteEditMetrics.propTypes = {
  dispatch: React.PropTypes.func.isRequired,
  error: React.PropTypes.string,
  interimNote: React.PropTypes.shape({
    get: React.PropTypes.func.isRequired,
  }).isRequired,
};

module.exports = NoteEditMetrics;
