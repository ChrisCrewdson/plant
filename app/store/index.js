const { createStore, applyMiddleware } = require('redux');
const reducers = require('../reducers'); // combineReducers already called on reducers in her)e
const api = require('../middleware/api');
const { setupSubscribe: userSubscribe } = require('./user');
const Immutable = require('immutable');

let middleware = [api];

if (process.env.NODE_ENV !== 'production') {
  const logger = require('../middleware/logger');
  middleware = [logger, api];
}

// Add the api to the pipeline/chain
const createStoreWithMiddleware = applyMiddleware(...middleware)(createStore);

const store = createStoreWithMiddleware(reducers, Immutable.fromJS(window.__INITIAL_STATE__ || {}));

userSubscribe(store);

module.exports = store;
