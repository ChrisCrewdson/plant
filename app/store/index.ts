export {}; // To get around: Cannot redeclare block-scoped variable .ts(2451)

const { createStore, applyMiddleware } = require('redux');

const seamless = require('seamless-immutable').static;
const reducers = require('../reducers'); // combineReducers already called on reducers in her)e
const api = require('../middleware/api');
const { setupSubscribe: userSubscribe } = require('./user');

const middleware = [api];

if (process.env.NODE_ENV !== 'production') {
  // eslint-disable-next-line global-require
  const logger = require('../middleware/logger');
  middleware.unshift(logger);
}

// Add the api to the pipeline/chain
const createStoreWithMiddleware = applyMiddleware(...middleware)(createStore);

// @ts-ignore - __INITIAL_STATE__ is added on the server
const initialState = seamless.from(window.__INITIAL_STATE__ || {});
const store = createStoreWithMiddleware(reducers, initialState);

userSubscribe(store);

module.exports = store;
