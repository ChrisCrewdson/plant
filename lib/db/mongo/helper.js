const _ = require('lodash');

module.exports = class Helper {
  static removeEmpty(doc) {
    if (!doc) {
      return doc;
    }
    if (_.isArray(doc)) {
      return doc.map(item => Helper.removeEmpty(item));
    }
    return _.pickBy(doc, item => item !== '');
  }

  /**
   * Helper function to convert _id from MongoId to string. Used in reads
   * @param {object} obj - Object that might have an _id
   * @returns {object} - the same object with a converted _id
   */
  static convertIdToString(obj) {
    if (_.isArray(obj)) {
      return obj.map(Helper.convertIdToString);
    }
    if (obj && obj._id) {
      // eslint-disable-next-line no-param-reassign
      obj._id = obj._id.toString();
    }
    return obj;
  }
};
