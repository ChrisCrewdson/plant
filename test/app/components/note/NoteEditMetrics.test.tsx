import React from 'react';
import renderer from 'react-test-renderer';
import { MemoryRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import { Dispatch } from 'redux';

import { MuiThemeProvider } from '@material-ui/core/styles';

import NoteEditMetrics from '../../../../app/components/note/NoteEditMetrics';
import store from '../../../../app/store';
import App from '../../../../app/components/App';
import { PlantAction } from '../../../../lib/types/redux-payloads';
import { theme } from '../../../helper';
import { UiInterimNote } from '../../../../lib/db/mongo/model-note';

describe('NoteEditMetrics', () => {
  const storeDispatch = store.dispatch;

  beforeAll(() => {
    store.dispatch = (() => {}) as Dispatch<PlantAction<any>>;
  });

  afterAll(() => {
    store.dispatch = storeDispatch;
  });

  test('should render NoteEditMetrics', () => {
    const interimNote = {
    } as UiInterimNote;

    const component = renderer.create(
      <MuiThemeProvider theme={theme}>
        <Provider store={store}>
          <App>
            <MemoryRouter>
              <NoteEditMetrics
                dispatch={storeDispatch}
                interimNote={interimNote}
              />
            </MemoryRouter>
          </App>
        </Provider>
      </MuiThemeProvider>);
    const noteEditMetrics = component.toJSON();
    expect(noteEditMetrics).toMatchSnapshot();
  });
});
