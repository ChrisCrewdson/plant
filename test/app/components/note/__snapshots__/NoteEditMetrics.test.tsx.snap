// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NoteEditMetrics should render NoteEditMetrics 1`] = `
<div
  className="react-root"
>
  <div
    style={
      Object {
        "textAlign": "left",
      }
    }
  >
    <div>
      <div
        className="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
        style={
          Object {
            "marginLeft": 20,
          }
        }
      >
        <label
          className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
          data-shrink={false}
          htmlFor="height"
          id="height-label"
          style={
            Object {
              "fontSize": "14px",
            }
          }
        >
          Height (inches only)
        </label>
        <div
          className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl"
          onClick={[Function]}
          style={
            Object {
              "fontSize": "14px",
            }
          }
        >
          <input
            aria-invalid={false}
            className="MuiInputBase-input MuiInput-input"
            disabled={false}
            id="height"
            name="height"
            onAnimationStart={[Function]}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            placeholder="Enter height of plant"
            required={false}
            type="number"
            value=""
          />
        </div>
        
      </div>
      <div
        className="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
        style={
          Object {
            "marginLeft": 20,
          }
        }
      >
        <label
          className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
          data-shrink={false}
          htmlFor="girth"
          id="girth-label"
          style={
            Object {
              "fontSize": "14px",
            }
          }
        >
          Girth (inches only)
        </label>
        <div
          className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl"
          onClick={[Function]}
          style={
            Object {
              "fontSize": "14px",
            }
          }
        >
          <input
            aria-invalid={false}
            className="MuiInputBase-input MuiInput-input"
            disabled={false}
            id="girth"
            name="girth"
            onAnimationStart={[Function]}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            placeholder="Enter girth of plant"
            required={false}
            type="number"
            value=""
          />
        </div>
        
      </div>
      <div
        className="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
        style={
          Object {
            "marginLeft": 20,
          }
        }
      >
        <label
          className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
          data-shrink={false}
          htmlFor="harvestCount"
          id="harvestCount-label"
          style={
            Object {
              "fontSize": "14px",
            }
          }
        >
          Harvest Count
        </label>
        <div
          className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl"
          onClick={[Function]}
          style={
            Object {
              "fontSize": "14px",
            }
          }
        >
          <input
            aria-invalid={false}
            className="MuiInputBase-input MuiInput-input"
            disabled={false}
            id="harvestCount"
            name="harvestCount"
            onAnimationStart={[Function]}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            placeholder="Enter number of items harvested"
            required={false}
            type="number"
            value=""
          />
        </div>
        
      </div>
      <div
        className="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
        style={
          Object {
            "marginLeft": 20,
          }
        }
      >
        <label
          className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
          data-shrink={false}
          htmlFor="harvestWeight"
          id="harvestWeight-label"
          style={
            Object {
              "fontSize": "14px",
            }
          }
        >
          Harvest Weight (lbs only)
        </label>
        <div
          className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl"
          onClick={[Function]}
          style={
            Object {
              "fontSize": "14px",
            }
          }
        >
          <input
            aria-invalid={false}
            className="MuiInputBase-input MuiInput-input"
            disabled={false}
            id="harvestWeight"
            name="harvestWeight"
            onAnimationStart={[Function]}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            placeholder="Enter weight of harvest"
            required={false}
            type="number"
            value=""
          />
        </div>
        
      </div>
      <label
        className="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementStart"
        style={
          Object {
            "display": "block",
          }
        }
      >
        <span
          className="MuiSwitch-root"
        >
          <span
            aria-disabled={false}
            className="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-78 MuiSwitch-switchBase MuiSwitch-colorPrimary"
            onBlur={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            style={
              Object {
                "maxWidth": "200px",
                "paddingLeft": "5px",
              }
            }
            tabIndex={null}
          >
            <span
              className="MuiIconButton-label"
            >
              <input
                checked={false}
                className="PrivateSwitchBase-input-81 MuiSwitch-input"
                name="firstBlossom"
                onChange={[Function]}
                type="checkbox"
              />
              <span
                className="MuiSwitch-thumb"
              />
            </span>
            <span
              className="MuiTouchRipple-root"
            />
          </span>
          <span
            className="MuiSwitch-track"
          />
        </span>
        <span
          className="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
        >
          First Blossom
        </span>
      </label>
      <label
        className="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementStart"
        style={
          Object {
            "display": "block",
          }
        }
      >
        <span
          className="MuiSwitch-root"
        >
          <span
            aria-disabled={false}
            className="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-78 MuiSwitch-switchBase MuiSwitch-colorPrimary"
            onBlur={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            style={
              Object {
                "maxWidth": "200px",
                "paddingLeft": "5px",
              }
            }
            tabIndex={null}
          >
            <span
              className="MuiIconButton-label"
            >
              <input
                checked={false}
                className="PrivateSwitchBase-input-81 MuiSwitch-input"
                name="lastBlossom"
                onChange={[Function]}
                type="checkbox"
              />
              <span
                className="MuiSwitch-thumb"
              />
            </span>
            <span
              className="MuiTouchRipple-root"
            />
          </span>
          <span
            className="MuiSwitch-track"
          />
        </span>
        <span
          className="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
        >
          Last Blossom
        </span>
      </label>
      <label
        className="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementStart"
        style={
          Object {
            "display": "block",
          }
        }
      >
        <span
          className="MuiSwitch-root"
        >
          <span
            aria-disabled={false}
            className="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-78 MuiSwitch-switchBase MuiSwitch-colorPrimary"
            onBlur={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            style={
              Object {
                "maxWidth": "200px",
                "paddingLeft": "5px",
              }
            }
            tabIndex={null}
          >
            <span
              className="MuiIconButton-label"
            >
              <input
                checked={false}
                className="PrivateSwitchBase-input-81 MuiSwitch-input"
                name="firstBud"
                onChange={[Function]}
                type="checkbox"
              />
              <span
                className="MuiSwitch-thumb"
              />
            </span>
            <span
              className="MuiTouchRipple-root"
            />
          </span>
          <span
            className="MuiSwitch-track"
          />
        </span>
        <span
          className="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
        >
          First Bud
        </span>
      </label>
      <label
        className="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementStart"
        style={
          Object {
            "display": "block",
          }
        }
      >
        <span
          className="MuiSwitch-root"
        >
          <span
            aria-disabled={false}
            className="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-78 MuiSwitch-switchBase MuiSwitch-colorPrimary"
            onBlur={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            style={
              Object {
                "maxWidth": "200px",
                "paddingLeft": "5px",
              }
            }
            tabIndex={null}
          >
            <span
              className="MuiIconButton-label"
            >
              <input
                checked={false}
                className="PrivateSwitchBase-input-81 MuiSwitch-input"
                name="harvestStart"
                onChange={[Function]}
                type="checkbox"
              />
              <span
                className="MuiSwitch-thumb"
              />
            </span>
            <span
              className="MuiTouchRipple-root"
            />
          </span>
          <span
            className="MuiSwitch-track"
          />
        </span>
        <span
          className="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
        >
          Harvest Start
        </span>
      </label>
      <label
        className="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementStart"
        style={
          Object {
            "display": "block",
          }
        }
      >
        <span
          className="MuiSwitch-root"
        >
          <span
            aria-disabled={false}
            className="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-78 MuiSwitch-switchBase MuiSwitch-colorPrimary"
            onBlur={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            style={
              Object {
                "maxWidth": "200px",
                "paddingLeft": "5px",
              }
            }
            tabIndex={null}
          >
            <span
              className="MuiIconButton-label"
            >
              <input
                checked={false}
                className="PrivateSwitchBase-input-81 MuiSwitch-input"
                name="harvestEnd"
                onChange={[Function]}
                type="checkbox"
              />
              <span
                className="MuiSwitch-thumb"
              />
            </span>
            <span
              className="MuiTouchRipple-root"
            />
          </span>
          <span
            className="MuiSwitch-track"
          />
        </span>
        <span
          className="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
        >
          Harvest End
        </span>
      </label>
      <label
        className="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementStart"
        style={
          Object {
            "display": "block",
          }
        }
      >
        <span
          className="MuiSwitch-root"
        >
          <span
            aria-disabled={false}
            className="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-78 MuiSwitch-switchBase MuiSwitch-colorPrimary"
            onBlur={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            style={
              Object {
                "maxWidth": "200px",
                "paddingLeft": "5px",
              }
            }
            tabIndex={null}
          >
            <span
              className="MuiIconButton-label"
            >
              <input
                checked={false}
                className="PrivateSwitchBase-input-81 MuiSwitch-input"
                name="leafShedStart"
                onChange={[Function]}
                type="checkbox"
              />
              <span
                className="MuiSwitch-thumb"
              />
            </span>
            <span
              className="MuiTouchRipple-root"
            />
          </span>
          <span
            className="MuiSwitch-track"
          />
        </span>
        <span
          className="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
        >
          Leaf Shed Start
        </span>
      </label>
      <label
        className="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementStart"
        style={
          Object {
            "display": "block",
          }
        }
      >
        <span
          className="MuiSwitch-root"
        >
          <span
            aria-disabled={false}
            className="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-78 MuiSwitch-switchBase MuiSwitch-colorPrimary"
            onBlur={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            style={
              Object {
                "maxWidth": "200px",
                "paddingLeft": "5px",
              }
            }
            tabIndex={null}
          >
            <span
              className="MuiIconButton-label"
            >
              <input
                checked={false}
                className="PrivateSwitchBase-input-81 MuiSwitch-input"
                name="leafShedEnd"
                onChange={[Function]}
                type="checkbox"
              />
              <span
                className="MuiSwitch-thumb"
              />
            </span>
            <span
              className="MuiTouchRipple-root"
            />
          </span>
          <span
            className="MuiSwitch-track"
          />
        </span>
        <span
          className="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
        >
          Leaf Shed End
        </span>
      </label>
    </div>
  </div>
</div>
`;
